tabular/src/autogluon/tabular/models/abstract/abstract_model.py

```diff
@@ -201,14 +212,28 @@ def preprocess(self, X):
             self.features = list(X.columns)  # TODO: add fit and transform versions of preprocess instead of doing this
             ignored_type_group_raw = self.params_aux.get('ignored_type_group_raw', [])
             ignored_type_group_special = self.params_aux.get('ignored_type_group_special', [])
-            valid_features = self.feature_metadata.get_features(invalid_raw_types=ignored_type_group_raw, invalid_special_types=ignored_type_group_special)
+            # TODO: Consider changing how this works or where it is done
+            if self.feature_metadata is None:
+                feature_metadata = FeatureMetadata.from_df(X)
+            else:
+                feature_metadata = self.feature_metadata
+            valid_features = feature_metadata.get_features(invalid_raw_types=ignored_type_group_raw, invalid_special_types=ignored_type_group_special)
             self.features = [feature for feature in self.features if feature in valid_features]
             if not self.features:
                 raise NoValidFeatures
             if list(X.columns) != self.features:
                 X = X[self.features]
         return X
 
+    # TODO: Remove kwargs?
+    def _preprocess_stateful(self, X, **kwargs):
+        """
+        Data transformation logic that is stateful and fold specific should be added here.
```

Comment:
perhaps provide an example of this and specifically mention "logic that depends on the data values themselves" since 'stateful' is not very specific term to use in the documentation given it's already in the method name.

Reply:
also clarify "fold" here refers to each fold of k-fold bagging

Reply:
Rewrote